File: leetcode015.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        ls = len(nums)
        res = []
        for i in range(ls-2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j = i+1
            k = ls-1
            while j < k:

                if nums[i]+nums[j]+nums[k] == 0:
                    res.append([nums[i],nums[j],nums[k]])
                    while j < k and nums[j] == nums[j + 1]:
                        j += 1
                    while j < k and nums[k] == nums[k - 1]:
                        k -= 1
                    j += 1
                    k -= 1
                elif nums[i]+nums[j]+nums[k] > 0:
                    k = k - 1
                else:
                    j = j + 1
        return res

if __name__ == "__main__":
    s = Solution()
    print(s.threeSum([-2,0,1,1,2]))

File: leetcodec++/leetcode015.cpp
//
// Created by 黄敏辉 on 2019-06-26.
//

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        if(nums.size()<=2)return res;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 2; i++){
            int a = nums[i];
            if (a > 0)
                break;
            if (i > 0 && a == nums[i-1])
                continue;
            for (int j = i+1, k = nums.size() - 1; j < k;){
                int b = nums[j];
                int c = nums[k];
                int val = a + b + c;
                if (val == 0) {
                    res.push_back(vector<int>({a, b, c}));
                    while (j<k && b == nums[++j]);
                    while (j < k &&c == nums[--k]);
                } else if (val > 0)
                    k--;
                else
                    j++;
            }
        }
        return res;
    }
};